http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/internal/response/ResponseHandlerHelper.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.crt.internal.response;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.crt.http.HttpClientConnection;
+import software.amazon.awssdk.crt.http.HttpHeader;
+import software.amazon.awssdk.crt.http.HttpHeaderBlock;
+import software.amazon.awssdk.crt.http.HttpStream;
+import software.amazon.awssdk.http.HttpStatusFamily;
+import software.amazon.awssdk.http.SdkHttpResponse;
+
+/**
+ * This is the helper class that contains common logic shared between {@link CrtResponseAdapter} and
+ * {@link InputStreamAdaptingHttpStreamResponseHandler}.
+ *
+ * CRT connection will only be closed, i.e., not reused, in one of the following conditions:
+ * 1. 5xx server error OR
+ * 2. It fails to read the response.
+ */
+@SdkInternalApi
+public class ResponseHandlerHelper {
+
+    private final SdkHttpResponse.Builder responseBuilder;
+    private final HttpClientConnection connection;
+    private AtomicBoolean connectionClosed = new AtomicBoolean(false);
+
+    public ResponseHandlerHelper(SdkHttpResponse.Builder responseBuilder, HttpClientConnection connection) {
+        this.responseBuilder = responseBuilder;
+        this.connection = connection;
+    }
+
+    public void onResponseHeaders(HttpStream stream, int responseStatusCode, int headerType, HttpHeader[] nextHeaders) {
+        if (headerType == HttpHeaderBlock.MAIN.getValue()) {
+            for (HttpHeader h : nextHeaders) {
+                responseBuilder.appendHeader(h.getName(), h.getValue());
+            }
+            responseBuilder.statusCode(responseStatusCode);
+        }
+    }
+
+    /**
+     * Release the connection back to the pool so that it can be reused.
+     */
+    public void releaseCrtConnection(HttpStream stream) {
+        if (connectionClosed.compareAndSet(false, true)) {
+            connection.close();
+            stream.close();
+        }
+    }
+
+    /**
+     * Close the connection completely
+     */
+    public void closeCrtConnection(HttpStream stream) {
```

Comment:
I would suggest to add another bool to the args for shutdown the connection or not, instead of have two different method, which feels more natural to me.

Also, I would rename it to releaseCrtConnectionAndStream/releaseCrtHttpResources, since the passed-in stream also gets released.

Reply:
I actually like having two methods with meaningful names because it makes it more readable IMO. I'll definitely forget what passing true to `closeCrtResource(stream, true)` implies after a couple of days if I don't go to the definition of the method.

Reply:
I considered `closeCrtResource`, but don't feel it's as clear as just `closeConnection` since crt stream and crt connection are just implementation details and they are part of a connection.